lib/cucumber/tcl/framework.tcl

```diff
@@ -12,31 +12,42 @@ set STEPS [list]
 #
 # Define procs to match Gherkin keyworkds that put data in the STEPS array
 #
-proc Given {re body} {
-  global STEPS
-  lappend STEPS [list $re $body]
+proc Given args {
+  _add_step {*}$args
 }
 
-proc When {re body} {
-  global STEPS
-  lappend STEPS [list $re $body]
+proc When args {
+  _add_step {*}$args
 }
 
-proc Then {re body} {
-  global STEPS
-  lappend STEPS [list $re $body]
+proc Then args {
+  _add_step {*}$args
 }
 
-proc And {re body} {
+proc And args {
+  _add_step {*}$args
+}
+
+proc _add_step args {
+
   global STEPS
-  lappend STEPS [list $re $body]
+
+  if {[llength $args] == 2} {
+    set re [lindex $args 0]
+    set params {}
+    set body [lindex $args 1]
+  } elseif {[llength $args] == 3} {
+    set re [lindex $args 0]
+    set params [lindex $args 1]
+    set body [lindex $args 2]
+  }
```

Comment:
The way I've written this was that you can either pass in 2 parameters (i.e. a regexp and a body) or 3 parameters (i.e. regexp, list of parameters and a body), so in the case of a regexp that can capture multiple matches, the Given call would look something like:

    Given {^Match1 (\w*), Match2 (\d*), Match 3 (\w*)} {match1 match2 match3} {
      do_something
    }

Do you think it would be better to have a separate parameter for each match? Or does this look ok (maybe with an error raised if the number of parameters is incorrect?)

I've just realised that I need to updatethe README to reflect this, but will hold off in case we decide to change this.